core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/protocol/query/QueryProtocolMarshaller.java

```diff
@@ -15,12 +15,147 @@
 
 package software.amazon.awssdk.core.internal.protocol.query;
 
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicInteger;
 import software.amazon.awssdk.annotations.SdkProtectedApi;
-import software.amazon.awssdk.core.SdkRequest;
-import software.amazon.awssdk.core.protocol.ProtocolRequestMarshaller;
+import software.amazon.awssdk.core.DefaultRequest;
+import software.amazon.awssdk.core.Request;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.core.protocol.MarshallingType;
+import software.amazon.awssdk.core.protocol.OperationInfo;
+import software.amazon.awssdk.core.protocol.ProtocolMarshaller;
+import software.amazon.awssdk.core.protocol.SdkField;
+import software.amazon.awssdk.core.protocol.SdkPojo;
+import software.amazon.awssdk.core.protocol.traits.DefaultValueTrait;
+import software.amazon.awssdk.core.protocol.traits.ListTrait;
+import software.amazon.awssdk.core.protocol.traits.MapTrait;
+import software.amazon.awssdk.core.util.SdkAutoConstructList;
+import software.amazon.awssdk.core.util.UriResourcePathUtils;
 
+/**
+ * Implementation of {@link ProtocolMarshaller} for AWS Query services.
+ *
+ * @param <OrigRequestT> Type of the original request object.
+ */
 @SdkProtectedApi
-public class QueryProtocolMarshaller<OrigRequestT extends SdkRequest>
-    implements ProtocolRequestMarshaller<OrigRequestT> {
+public final class QueryProtocolMarshaller<OrigRequestT>
+    implements ProtocolMarshaller<Request<OrigRequestT>> {
+
+    private static final QueryMarshallerRegistry MARSHALLER_REGISTRY = QueryMarshallerRegistry
+        .builder()
+        .marshaller(MarshallingType.STRING, SimpleTypeQueryMarshaller.STRING)
+        .marshaller(MarshallingType.INTEGER, SimpleTypeQueryMarshaller.INTEGER)
+        .marshaller(MarshallingType.FLOAT, SimpleTypeQueryMarshaller.FLOAT)
+        .marshaller(MarshallingType.BOOLEAN, SimpleTypeQueryMarshaller.BOOLEAN)
+        .marshaller(MarshallingType.DOUBLE, SimpleTypeQueryMarshaller.DOUBLE)
+        .marshaller(MarshallingType.LONG, SimpleTypeQueryMarshaller.LONG)
+        .marshaller(MarshallingType.INSTANT, SimpleTypeQueryMarshaller.INSTANT)
+        .marshaller(MarshallingType.SDK_BYTES, SimpleTypeQueryMarshaller.SDK_BYTES)
+        .marshaller(MarshallingType.NULL, SimpleTypeQueryMarshaller.NULL)
+        .marshaller(MarshallingType.MAP, new MapMarshaller())
+        .marshaller(MarshallingType.LIST, new ListMarshaller())
+        .marshaller(MarshallingType.SDK_POJO, (req, path, val, sdkField) -> doMarshall(path, req, val))
```

Comment:
Do we need a BigDecimal marshalling type?

Reply:
Nah we have that as a customization for budgets. We actually don't support the "true" big decimal type as defined in the aws service framework. It's modeled as a string in budgets.